### PR TITLE
fix loss masking and padding

### PR DESCRIPTION
**Pull request by sohamparikh**
# ✨ Description

Fixes and improvements for loss masking and padding

Closes # 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

- Discard masked tokens while computing loss mean
- Mask -ve tokens in embeddings
- Fix cached yaml comparison when `truncate_documents=False`
- sum(long_docs_filter) was extremely slow for large datasets

### CODE REVIEW

**Review by tobyzl2 — APPROVED** [review]

File: fast_llm/data/dataset/gpt/sampled.py [hunk 1]
```diff
@@ -525,8 +529,8 @@ def _load_yaml_data(self, data: dict[str, typing.Any]) -> None:
         elif "unshuffled_tokens" not in data:
             # Backward compatibility
             # TODO v0.x: Remove
-            assert self._truncate_documents
-            data["unshuffled_tokens"] = data["tokens_per_epoch"] * data["unshuffled_epochs"]
+            assert not self._truncate_documents
+            data["unshuffled_tokens"] = data["dataset"]["tokens_per_epoch"] * data["unshuffled_epochs"]
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py]: yes, i got a bit confused about the purpose of this. There's still an issue with `yaml_data` not containing `unshuffled_tokens` (we can't get it without building the padded cumsum). Pushed a hack to copy it from `loaded_yaml_data` instead to avoid breaking the flow, not sure if there's a cleaner way

File: fast_llm/functional/cross_entropy.py [hunk 2]
```diff
@@ -145,7 +145,7 @@ def _fused_cross_entropy_forward_backward(
 
     per_sample_loss = sum_exp_logits.log() - predicted_logits
     if loss_mask is not None:
-        per_sample_loss = per_sample_loss * loss_mask
+        per_sample_loss = per_sample_loss[loss_mask]
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/functional/cross_entropy.py:148]: The loss is not as interpretable/comparable when we include loss from masked tokens (0) in the average. We start seeing a lot of variance in loss when mixing samples with/without masked tokens

File: fast_llm/layers/language_model/embedding.py [hunk 3]
```diff
@@ -99,7 +99,10 @@ def _forward(self, input_: torch.Tensor, position_ids: torch.Tensor | None) -> t
                 input_ = split(input_, group=group, dim=0)
                 if self._use_absolute_position_embeddings:
                     position_ids = split(position_ids, group=group, dim=0)
-            embeddings = torch.embedding(self.word_embeddings_weight, input_)
+            # mask padded tokens
+            input_mask = input_ >= 0
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/layers/language_model/embedding.py:103]: We set the padded tokens to -100 mainly to mask loss on them. Many tokenizers don't have pad tokens either so not straightforward to take it from the config either

File: fast_llm/functional/cross_entropy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/functional/cross_entropy.py:148]: Oh, so it's to have the right denominator in the mean? Indexing is a bad idea because it introduces a cuda synchronization point (really slow), but you can divide by `loss_mask.sum()` instead.

Also we probably want to deal with the case `loss_mask.sum()==0`  .

File: fast_llm/layers/language_model/embedding.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/layers/language_model/embedding.py:103]: We can add a flag, something like we already do for labels https://github.com/ServiceNow/Fast-LLM/blob/main/fast_llm/models/gpt/model.py#L332.
Also I just noticed the method is within `@torch.compile`, so the overhead shouldn't be too noticeable.

File: fast_llm/data/data/gpt/config.py [hunk 4]
```diff
@@ -48,14 +48,15 @@ class GPTDataConfig(DataConfig, GPTLegacyConfig):
         desc="Multiprocessing context. Do not touch.",
         hint=FieldHint.expert,
     )
-    truncate_documents: bool = Field(
-        default=True,
+    truncate_documents: bool | None = Field(
```
Comment (jlamypoirier) [inline fast_llm/data/data/gpt/config.py:51]: That works, but we normally do backward compatibility in `_from_dict`, see example in lines 73-90 below. This one can go in `GPTTrainerConfig._from_dict`.
Also needs a todo for removal.

**Review by jlamypoirier — APPROVED** [review]
